onmt/translate/BeamSearcher.lua

```diff
@@ -0,0 +1,298 @@
+--[[ BeamSearcher is a class used for performing general beam search.
+--]]
+local BeamSearcher = torch.class('BeamSearcher')
+
+local function recursiveApply(h, func, ...)
+  local hOut
+  if torch.type(h) == 'table' then
+    hOut = {}
+    for key, val in pairs(h) do
+      hOut[key] = recursiveApply(val, func, ...)
+    end
+    return hOut
+  end
+  if torch.isTensor(h) then
+    hOut = func(h, ...)
+  else
+    hOut = h
+  end
+  return hOut
+end
+
+local function recursiveClone(hIn)
+  return recursiveApply(hIn, function (h) return h:clone() end)
+end
+
+local function beamReplicate(hIn, beamSizeIn)
+  return recursiveApply(hIn, function (h, beamSize)
+    local batchSize = h:size(1)
+    local sizes = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    return h:contiguous():view(batchSize, 1, table.unpack(sizes)):expand(batchSize, beamSize, table.unpack(sizes)):contiguous():view(batchSize * beamSize, table.unpack(sizes))
+  end, beamSizeIn)
+end
+
+local function beamSelect(hIn, selIndexesIn)
+  return recursiveApply(hIn, function (h, selIndexes)
+    local batchSize = selIndexes:size(1)
+    local beamSize = selIndexes:size(2)
+    return h:index(1, selIndexes:view(-1):long() + (torch.range(0, (batchSize - 1) * beamSize, beamSize):long()):contiguous():view(batchSize, 1):expand(batchSize, beamSize):contiguous():view(-1))
+  end, selIndexesIn)
+end
+
+local function flatToRc(hIn, beamSizeIn)
+  return recursiveApply(hIn, function (h, beamSize)
+    local batchSize = math.floor(h:size(1) / beamSize)
+    local sizes = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    return h:view(batchSize, beamSize, table.unpack(sizes))
+  end, beamSizeIn)
+end
+
+local function selectBatch(hIn, remainingIn)
+  return recursiveApply(hIn, function (h, remaining)
+    if not torch.isTensor(remaining) then
+      remaining = torch.LongTensor(remaining)
+    end
+    return h:index(1, remaining)
+  end, remainingIn)
+end
+
+local function selectBatchBeam(hIn, beamSizeIn, batchIn, beamIn)
+  return recursiveApply(hIn, function (h, beamSize, batch, beam)
+    local batchSize = math.floor(h:size(1) / beamSize)
+    local sizes = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    local hOut = h:view(batchSize, beamSize, table.unpack(sizes))
+    return hOut[{batch, beam}]
+  end, beamSizeIn, batchIn, beamIn)
+end
+
+local function rcToFlat(hIn)
+  return recursiveApply(hIn, function (h)
+    local sizes = {}
+    sizes[1] = h:size(1) * h:size(2)
+    for j = 3, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    return h:view(table.unpack(sizes))
+  end)
+end
+
+--[[ Construct a BeamSearcher object. BeamSearcher takes charge of the beam search related logic such as ranking the predictions, tracking the parents in each beam, and transforming the outputs from last time step according to the parent indexes in the current time step.
```

Comment:
ok I fixed most, but I'm not sure how to break the parameter explanations into multiple lines?